gossip/src/contact_info.rs

```diff
@@ -0,0 +1,293 @@
+    }
+    let addr = socket.ip();
+    if addr.is_unspecified() {
+        return Err(Error::UnspecifiedAddr(addr));
+    }
+    if addr.is_multicast() {
+        return Err(Error::MulticastAddr(addr));
+    }
+    Ok(())
+}
+
+impl Sanitize for ContactInfo {
+    fn sanitize(&self) -> Result<(), SanitizeError> {
+        if self.wallclock >= MAX_WALLCLOCK {
+            return Err(SanitizeError::ValueOutOfBounds);
+        }
+        Ok(())
+    }
+}
+
+impl Serialize for ContactInfo {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        ContactInfoLite {
+            pubkey: Cow::Borrowed(&self.pubkey),
+            wallclock: self.wallclock,
+            shred_version: self.shred_version,
+            version: Cow::Borrowed(&self.version),
+            addrs: Cow::Borrowed(&self.addrs),
+            sockets: Cow::Borrowed(&self.sockets),
+        }
+        .serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for ContactInfo {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let ContactInfoLite {
+            pubkey,
+            wallclock,
+            shred_version,
+            version,
+            addrs,
+            sockets,
+        } = ContactInfoLite::deserialize(deserializer)?;
+        // TODO: sanity check on the fields.
+        // TODO: verify port offsets don't overflow.
+        let mut node = ContactInfo {
+            pubkey: pubkey.into_owned(),
+            wallclock,
+            shred_version,
+            version: version.into_owned(),
+            addrs: addrs.into_owned(),
+            sockets: sockets.into_owned(),
+            ..ContactInfo::default()
+        };
+        node.cache = std::array::from_fn(|k| node.get_socket(k as u8));
+        Ok(node)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        rand::{seq::SliceRandom, Rng},
+        std::{
+            collections::{HashMap, HashSet},
+            iter::repeat_with,
+            net::{Ipv4Addr, Ipv6Addr},
+            ops::Range,
+        },
+    };
+
+    fn new_rand_addr<R: Rng>(rng: &mut R) -> IpAddr {
+        if rng.gen() {
+            let addr = Ipv4Addr::new(rng.gen(), rng.gen(), rng.gen(), rng.gen());
+            IpAddr::V4(addr)
+        } else {
+            let addr = Ipv6Addr::new(
+                rng.gen(),
+                rng.gen(),
+                rng.gen(),
+                rng.gen(),
+                rng.gen(),
+                rng.gen(),
+                rng.gen(),
+                rng.gen(),
+            );
+            IpAddr::V6(addr)
+        }
+    }
+
+    fn new_rand_port<R: Rng>(rng: &mut R) -> u16 {
+        let port = rng.gen::<u16>();
+        let bits = u16::BITS - port.leading_zeros();
+        let shift = rng.gen_range(0u32, bits + 1u32);
+        port.checked_shr(shift).unwrap_or_default()
+    }
```

Comment:
I don't understand what change you are suggesting.
This occasionally creates invalid socket addresses which is intentional to have test coverage over paths which fail `sanitize_socket`.

If we want to make the test deterministic we can use a `SeedableRng` and a seeded random number generator.
`Pubkey::new_unique` approach isn't good because:
https://discord.com/channels/428295358100013066/439194979856809985/881943239823880222

Reply:
Even if all corner cases are identified, it is impossible to cover all their various combinations using deterministic tests. Using unseeded random number generator here, we are getting a much larger test coverage.

Flaky tests by themselves are not a problem, as long as they are easily reproducible locally; in which case they do identify a legit bug or corner case.

Given that this is a single threaded test and there is no other dependence on runtime environment here, do you see any reason that if this test becomes flaky it won't be locally reproducible, by say running the test 1000 times in a loop?

Reply:
how about switch to deterministic rng, seed with time and log the seed? so we can immediately reproduce a failure with trivial modification

Reply:
you didn't answer my question